bootstrap/src/main/java/jenkins/bootstrap/Plugin.java

```diff
@@ -0,0 +1,188 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.bootstrap;
+
+import org.apache.tools.ant.Project;
+import org.apache.tools.ant.taskdefs.Delete;
+import org.apache.tools.ant.taskdefs.Expand;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URL;
+import java.util.List;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import java.util.logging.Logger;
+import java.util.zip.ZipEntry;
+
+/**
+ * Plugin that we process during the bootstrap.
+ *
+ * <p>
+ * In order to identify the core component plugin that overrides Jenkins core and put them
+ * into the classpath, the bootstrap logic needs to inspect plugins lightly.
+ * Jenkins will later do further full treatment of plugins.
+ *
+ * <p>
+ * Some of the capabilities of the full plugin manager is missing from this code. For example,
+ * the bootstrap doesn't handle {@code *.jpl} file used during the development.
+ *
+ * @author Kohsuke Kawaguchi
+ */
+class Plugin {
+    /**
+     * foo.jpi file
+     */
+    private final File archive;
+    /**
+     * Directory to explode the plugin.
+     */
+    private final File destDir;
+
+    Plugin(File archive) {
+        this.archive = archive;
+        this.destDir = getDestDir();
+    }
+
+    void process(DependenciesTxt core, List<URL> jars, List<DependenciesTxt> overrides) throws IOException {
+        if (!isCoreComponentOverride() || isDisabled())
+            return;
+
+        explode();
+
+        File dependencies = new File(destDir,"META-INF/dependencies.txt");
+        try (InputStream i = new FileInputStream(dependencies)) {
+            DependenciesTxt deps = new DependenciesTxt(i);
+            for (Dependency d : deps.dependencies) {
+                if (core.hasNewerThan(d)) {
+                    // core has a newer version. abort.
+                    LOGGER.warning("Skipping core override plugin "+archive+" because core contains a newer version of "+d.ga);
+                    return;
+                }
+            }
+            overrides.add(deps);
+        }
+
+
+        File lib = new File(destDir, "WEB-INF/lib");
+        LOGGER.info("Loading core overrides from "+lib);
```

Comment:
This needs to be higher severity than the boring message above, and be more specific (perhaps elsewhere in the code). We need to know what actually gets loaded.